test-d/index.test-d.ts

```diff
@@ -0,0 +1,63 @@
+import {expectType} from 'tsd';
+import mem = require('../dist');
+
+const fn = (_string: string) => true;
+
+expectType<typeof fn>(mem(fn));
+expectType<typeof fn>(mem(fn, {maxAge: 1}));
+expectType<typeof fn>(mem(fn, {cacheKey: (...arguments_) => arguments_}));
+expectType<typeof fn>(
+	mem(
+		fn,
+		{
+			cacheKey: arguments_ => arguments_[0],
+			cache: new Map<string, {data: boolean; maxAge: number}>()
+		}
+	)
+);
+expectType<typeof fn>(
+	mem(fn, {cache: new Map<string, {data: boolean; maxAge: number}>()})
+);
+
+// Testing that the full cache object works with type inference
+function objectFnReturnsNumber(_args: {key: string}) {
+	return 10;
+}
+
+expectType<typeof objectFnReturnsNumber>(
+	mem(
+		objectFnReturnsNumber,
+		{
+			cacheKey: (_args: [{key: string}]) => {
+				return new Date();
+			},
+			cache: {
+				get: (_key: Date) => {
+					return {
+						data: 5,
+						maxAge: 2
+					};
+				},
+				set: (_key: Date, _value: { data: number; maxAge: number }) => {},
+				has: (_key: Date) => {
+					return true;
+				},
+				delete: (_key: Date) => {},
+				clear: () => {}
+			}
+		}
+	)
+);
```

Comment:
Since so many things were changed at once in #64 and #66 it's best we split these PRs. Moving files around especially makes it hard to see what changed exactly.

Do you mind sending this `cacheKey` test separately? I'll let you know when you can rebase this PR